Feature/resource channel

sitepark-veltrup opened the pull request:
A resource channel is a closed area in which the IES publishes its pages. This can be a live or preview area, for example.
The `ResourceChannel` is required to be able to read publication area-specific values. This can be the anchor of the publication area, for example.

Multi-language support is also included in this pull request.

sitepark-schaeper commented on src/ResourceChannel.php:
```diff
@@ -0,0 +1,34 @@
+ * into different channels. This can be the live channel for the
+ * website, but also a preview or intranet channel, for example.
+ *
+ * These channels have certain properties that are mapped in this class.
+ * are mapped in this class.
```
```suggestion
 * These channels have certain properties that are mapped in this class.
```

sitepark-schaeper commented on src/ResourceChannel.php:
```diff
@@ -0,0 +1,34 @@
+        public readonly string $nature,
+        public readonly string $locale,
+        public readonly string $encoding,
+        public readonly string $searchIndex,
```
why do we keep track of a "searchIndex" here - would that not be the `atoolo-search`s task?

sitepark-schaeper replied:
okay, if you say so :)

sitepark-schaeper commented on src/Loader/SiteKitLoader.php:
```diff
@@ -10,58 +10,87 @@
     public function __construct(
-        private readonly ResourceBaseLocator $baseLocator
+        private readonly ResourceBaseLocator $baseLocator,
+        private readonly ResourceChannelFactory $resourceChannel,
```
I think we should either rename this to `$resourceChannelFactory` or pass the `ResourceChannel` itself as it will always be `create`d exactly once upon first usage.

sitepark-schaeper commented on src/Loader/SiteKitLoader.php:
```diff
@@ -10,58 +10,87 @@
 class SiteKitLoader implements ResourceLoader
 {
+    /**
+     * @var array<string, string>|null
```
```suggestion
     * @var ?array<string, string> $langLocaleMap
```

sitepark-schaeper commented on src/ResourceHierarchyWalker.php:
```diff
@@ -414,12 +414,13 @@ public function next(): ?Resource
+    public function walk(Resource|string $base, callable $fn): void
     {
         $this->init($base);
         $fn($base);
```
```suggestion

        $fn($this->current);
```

sitepark-schaeper commented on src/ResourceChannel.php:
```diff
@@ -0,0 +1,34 @@
+        public readonly string $name,
+        public readonly string $anchor,
+        public readonly string $serverName,
+        public readonly bool $preview,
```
do we need a boolean for this? do we need this at all?  
at the very least it should be `isPreview`

sitepark-schaeper commented on src/ResourceChannel.php:
```diff
@@ -0,0 +1,34 @@
+        public readonly bool $preview,
+        public readonly string $nature,
+        public readonly string $locale,
+        public readonly string $encoding,
```
what do we need this for?

sitepark-schaeper replied:
okay, if you think it might be usefull in the future we can keep it.

sitepark-schaeper commented on src/Loader/SiteKitLoader.php:
```diff
@@ -10,58 +10,87 @@
 
+    private function locationToFile(string $location, string $lang): string
+    {
+        $file = $this->baseLocator->locate() . '/' . $location;
```
can the base change between invokations? I'd call `locate` once in the constructor and use the result here.  
I know the `ServerVarResourceBaseLocator` "caches" it's result (which I'd also argue it shouldn't) but not all implementations will do that.

sitepark-schaeper commented on src/Loader/SiteKitLoader.php:
```diff
@@ -112,12 +142,41 @@ private function loadRaw(string $location): array
+            return $lang;
+        }
+
+        if ($this->langLocaleMap === null) {
+            $this->langLocaleMap = $this->createLangLocaleMap();
+        }
```
```suggestion
        $this->langLocaleMap ??= $this->createLangLocaleMap();
```

sitepark-schaeper commented on src/Loader/ServerVarResourceBaseLocator.php:
```diff
@@ -5,27 +5,44 @@
 
-        if ($this->subDirectory === null) {
+        $resourceLayoutResourceBase = $_SERVER[$this->variableName] .
+            '/objects';
+        if (is_dir($resourceLayoutResourceBase)) {
+            $this->resourceBase = $resourceLayoutResourceBase;
+            return $resourceLayoutResourceBase;
+        }
+
+        if (is_dir($_SERVER[$this->variableName])) {
+            $this->resourceBase = $_SERVER[$this->variableName];
             return $_SERVER[$this->variableName];
         }
 
-        return $_SERVER[$this->variableName] . '/' . $this->subDirectory;
+        throw new RuntimeException(
+            "Resource root directory not found: " .
+            $_SERVER[$this->variableName]
+        );
```
I think we can make this a bit easier to understand:

```suggestion

        $base = $_SERVER[$this->variableName];

        $baseObjects = $base . DIRECTORY_SEPARATOR . 'objects';
        if (is_dir($baseObjects)) {
            return $this->resourceBase = $baseObjects;
        }

        if (is_dir($base)) {
            return $this->resourceBase = $base;
        }

        throw new RuntimeException(
            'Resource root directory not found: ' . $base
        );

```

sitepark-schaeper commented on src/SiteKitResourceChannelFactory.php:
```diff
@@ -0,0 +1,94 @@
+    public function create(): ResourceChannel
+    {
+
+        if ($this->resourceChannel !== null) {
+            return $this->resourceChannel;
+        }
```
shouldn't a `Factory` technically be able to create more than one instance?

sitepark-schaeper replied:
okay.  

We also don't have the case that we would want to create a new `ResourceChannel` because the `contenxt.php` might have changed or similar?

sitepark-schaeper commented on src/SiteKitResourceChannelFactory.php:
```diff
@@ -0,0 +1,94 @@
+
+        /** @var ContextPhp $context */
+        return $context;
+    }
+    private function findContextPhpFile(): string
```
```suggestion
    }

    private function findContextPhpFile(): string
```

sitepark-schaeper commented on src/SiteKitResourceChannelFactory.php:
```diff
@@ -0,0 +1,94 @@
+    }
+    private function findContextPhpFile(): string
+    {
+        $resourceBase = $this->resourceBaseLocator->locate();
+        if (file_exists($resourceBase . '/../context.php')) {
+            $contextPhpFile = dirname($resourceBase) . '/context.php';
+        } else {
+            $contextPhpFile = $resourceBase . '/WEB-IES/context.php';
+        }
```
it seems really weird to specifically require a class that locates a certain directory and to then check for a file outside of it.

sitepark-schaeper replied:
well yes, `dirname('/resource/base/')` still is `/resource`. So regardless of how we put it we still check for a file outside of the directory we've searched, right?

sitepark-schaeper reviewed (commented):
the `lang`, `language` and `locale` namings are very confusing to me. is it, that a lang/language is just "de" or "en" and a locale would be "de_DE"/"en_US"? or can they also include an encoding lik "en_EN.UTF-8"?
either way, I believe this should be cleaned up and made clear, as currently everything is just a string (some of which even default to `''`).

I've seen a couple of occurences of using `"/"` in paths - did you consider using `DIRECTORY_SEPARATOR`? I don't really know which one is better - This will probably never be executed on windows and even if, for file paths `"/"` should work too. Is there a specific reason you chose to do it this way?

sitepark-schaeper commented on src/Loader/SiteKitLoader.php:
```diff
@@ -10,58 +10,87 @@
 
         $init = $data['init'];
 
+        $locale = $init['locale'];
+        $resourceLang = substr($locale, 0, 2);
```
okay so - and correct me if im wrong:
- we load/search resources via either a locale (`de_DE`, `de_DE.UTF-8`, `zh_Hant-TW`), a language (`de`) or an empty string (but call it `$lang`)
- we aggregate locales in a specific format on resources (`de_DE`)
- and we use languages for the `Resource` class (`de`)

... and all of those are just strings. This is very confusing - especially if we start manipulating them with `substr` and alike.

sitepark-schaeper commented on src/ResourceLoader.php:
```diff
@@ -17,7 +17,7 @@ interface ResourceLoader
      * @throws ResourceNotFoundException
      */
-    public function load(string $location): Resource;
+    public function load(string $location, string $lang = ''): Resource;
 
-    public function exists(string $location): bool;
```
As a user that is given this signature I have no idea:
- what language do I get if I invoke `load($location)`?
- how does a `$lang` have to look like? `german`? `ger`? `de`? `de_DE`? `de-DE`? ...
- If I just want to load a resource and do not care about languages, can I omit `$lang` or do I have to find out what language the resource I'm trying to load is in?
- does a separate resource for a `$lang` have to exist somewhere? where exactly? or are they translated while loading?

String arguments or properties in general are often a big mistery as to what they (can) contain. Usualy they need extensive documentation to clarify that. If you want to go that route then we could reduce the places to put such a documentation if only a `ResourceLanguage` (or whatever) takes this argument instead of every single `ResourceLoader` method.

sitepark-schaeper reviewed (commented):
_necessary comment to be able to submit the code review_

sitepark-veltrup commented:
- I have now introduced a `ResourceLanguage` and `ResourceLocation` class
- `ResourceChannel` now contains `$resourceDir` and `$baseDir` and is passed directly to the `SiteKitLoader`
- The `ResourceBaseLocator` is no longer needed and has been removed.
- I took the opportunity to smooth out the `Resource` class and removed the getter methods
- The `init` values from the `Resource` are now placed directly in `DataBag` and no longer below `init`. ``$resource->data->getString('init.anchor')` was not so good.

sitepark-schaeper reviewed (approved):
Good stuff!
